### DISCUSSION TITLE
v4 release planning

### DISCUSSION THREAD

**Discussion by matthewhegarty** [django-import-export/django-import-export#1502] [Ideas]
Following the release of v3, we're now starting to think about v4.  I suggest to use this discussion to gather potential new features.

I've created a v4 label, and flagged the first PR: #1491

Any suggestions or contributions welcome.

**Comment by matthewhegarty** [comment]
#1376

There is some overlap between logic in import-export and DRF.  We could look at ways to reduce any duplication, and to optionally integrate DRF.  #1501 is a good example of where we could potentially be duplicating DRF logic.

**Comment by matthewhegarty** [comment]
#1452 

We seem to get a few requests about how to export a subset of data.  It is possible at the moment but not straightforward.  We may be able to make this process more user-friendly

**Comment by matthewhegarty** [comment]
#1440 

Export actions should support multiple resources in the changelist

**Comment by matthewhegarty** [comment]
#1459 

Allow for option to exclude dependencies (e.g. Tablib).

**Reply by matthewhegarty** [reply]
Additionally the xlwt package (dependency of tablib) appears to be unmaintained now.  Somehow we should allow usage of other libraries such as openpyxl

**Comment by matthewhegarty** [comment]
#1504  review deprecations for v4

**Comment by matthewhegarty** [comment]
#1369 Flake8 / Black checks

- Research and implement code quality tools
- Apply consistent formatting to existing code base

**Comment by matthewhegarty** [comment]
#445 handle inline data (lots of requests for this in the thread)

**Comment by matthewhegarty** [comment]
Option to disable 'confirmation' step on import.  This was kind of started in #1319 but never finished.  This could help with the recurring issue about non-sequential db ids (e.g. #560).

**Comment by matthewhegarty** [comment]
Define library behaviour in settings file (similar to other libraries like DRF).  At the moment settings are available on the Resource (via ResourceOptions) but this doesn't make sense when you want to provide library level config (for example, use of localization etc)

**Comment by matthewhegarty** [comment]
Better control over formatting of error messages in Admin UI (this comes up a lot in [threads](https://github.com/django-import-export/django-import-export/issues/1257#issuecomment-952276485))

**Comment by matthewhegarty** [comment]
Hi All

I'm starting to get together some PRs for a v4 release.  My plan is to try to mop up as many of the outstanding issues as I can and put them into a v4 release.  Any help with picking up issues, reviewing, comments etc is greatly appreciated.

Also, if you have any views for something you would like to see in v4, please reply in the thread.

I'm hoping to get this release out by the end of the year.

**Reply by andrewgy8** [reply]
Awesome work @matthewhegarty and thank you for driving this! 👏 

Something that I have been working on in the background is better documentation around the library. I would love to see this come with a v4 release.  Although it could happen in parallel, you have given me motivation and a self-inflicted deadline 😜

**Reply by matthewhegarty** [reply]
Hi Andrew

That sounds great.  I did give the docs a bit of a refresh in the last couple of releases but there's always room for improvement.  I'm currently working on this [PR](https://github.com/django-import-export/django-import-export/pull/1641) which is refactoring the interface for v4.  Mainly this is simply renaming methods and arg lists, so that they are standardised.  This will affect documentation, particularly the [import workflow](https://django-import-export.readthedocs.io/en/latest/import_workflow.html) page.  I was thinking I would rework that page at some point, possibly to include a sequence diagram which might clarify the workflow better.  You are welcome to help out if that is something you are interested in.

cheers
Matt

**Reply by matthewhegarty** [reply]
Have added a sequence diagram to #1641 - still a work in progress

**Comment by matthewhegarty** [comment]
Closing as thoughts turn to v5
